### PR TITLE
Differential privacy in janus

### PR DESCRIPTION
**Pull request by MxmUrw**
We have integrated differential privacy into janus as proposed here #1865.

Highlights:
 - The differential privacy strategy for a vdaf is carried directly in its enum variant in `VdafInstance` and `VdafOps`. This means that every vdaf is responsible to carry the data required for each of its supported strategies. No changes to database schema or task definition are required. For already existing types which do not need differential privacy the change is conservative - no updates to the enum variant have to occur. For a vdaf which supports multiple strategies, a dedicated enum containing *only* the supported dp strategies has to be created. An example for our type is [here](https://github.com/divviup/janus/blob/c22282e4777c01841350627662f3d51b2e62b050/core/src/task.rs#L25-L37).
 - We updated the `vdaf_dispatch!` and `vdaf_ops_dispatch!` macros to optionally generate the appropriate differential privacy strategy type and a value of that type from the data carried in `VdafInstance` and `VdafOps` respectively. For vdafs which don't support dp, the `NoDifferentialPrivacy` strategy is instantiated. For vdafs which support multiple strategies, multiple case-statements are generated - one for each possible strategy which it might be carrying.
 - We merged the three enum variants for the different bitsizes for our fixedvec type in `VdafInstance` into a single variant carrying the bitsize as data. For `VdafOps` this wasn't possible since it carries the already instantiated vdafs (as `Arc<A>`).

Note: We had to update prio a little bit, mostly deriving various traits for the new dp types. See this draft PR: https://github.com/divviup/libprio-rs/pull/751.

### CODE REVIEW

**Review by branlwyd — COMMENTED** [review]
This approach looks reasonable to me, just a few nitpicks. Once the related `libprio-rs` lands & is released I think this will be good to go.

File: aggregator/src/aggregator.rs
Comment (branlwyd) [reply aggregator/src/aggregator.rs:1171]: Still needs to be addressed.

File: core/src/dp.rs
Comment (branlwyd) [inline core/src/dp.rs:1]: nit: commented code

**Review by branlwyd — APPROVED** [review]
LGTM modulo a couple small nitpicks; can be merged once we get the libprio-rs release out.

File: docs/samples/tasks.yaml
Comment (divergentdave) [reply docs/samples/tasks.yaml:16]: Filed #1959

File: aggregator/src/aggregator.rs
Comment (MxmUrw) [reply aggregator/src/aggregator.rs:1171]: Yes, the problem here is that whether this variable is used depends on the given macro parameters. If `$dp_strategy` and `$DpStrategy` are given, then it is used. Otherwise it is not. This means that changing the name to `dp_strategy` results in lots of warnings, since in most invocations of `vdaf_ops` it is, in fact, not used.

Another solution to this problem would be to add `#[allow(unused_variables)]` before the match case but that would also suppress warnings about unused variables in the body. There are also some other solutions, with their own trade-offs, but I think using the underscore notation is the most straightforward solution.

Of course I'm open to other suggestions, in the meantime I'm going to add a short comment explaining the naming which I think should solve the situation for the most part.

File: messages/Cargo.toml
Comment (ooovi) [reply messages/Cargo.toml]: we're rebased to main, so this is done now! :)

**Comment by tgeoghegan** [comment]
@MxmUrw: sorry for neglecting this PR. We believe we're done (for now!) with big disruptive changes on Janus main, so I think we can land these changes without you needing to do any more annoying rebases. @divergentdave and I would both like to review -- if only so we are familiar with the changes -- but if Bran already signed off then I suspect there won't be any big blockers.

We're going to do our best to review in the next few days, so hopefully we can land this work later this week or sometime next. Several of us are traveling for work next week which might distract us a bit but I think we can get this done.

Thanks as ever for your patience.

**Comment by MxmUrw** [comment]
Thanks for the update! Looking forward to the review :)

**Review by divergentdave — COMMENTED** [review]
Looking good! I've got one error handling change to request, and then some suggestions for cleaning up the macros.

File: docs/samples/tasks.yaml
Comment (divergentdave) [inline docs/samples/tasks.yaml:17]: These parts of the file are ignored by `VdafInstance` deserialization. Note that for a task YAML file that uses Prio3FixedPointBoundedL2VecSum, the `dp_strategy` key would be nested inside the tagged map under the `vdaf` key, not at the top level of the task. We should roll back the changes to this file to avoid confusion.

When we eventually set up optional DP strategies for standardized VDAFs, we'll want to configure deserialization with defaults (of NoDifferentialPrivacy) for backwards compatibility with existing task YAML files, database JSON columns, and aggregator API JSON requests/responses.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
This fundamentally looks good, and I appreciate the simplification to dispatching the FixedPoint VDAFs. I have a few nits around error types and other things, but nothing that should be substantial blockers.

File: core/src/dp.rs
Comment (MxmUrw) [reply core/src/dp.rs]: Funnily enough it turns out that we don't need this enum anymore, since this functionality is handled on a per-vdaf-basis by the various enums in the `vdaf_instance_strategies` module in `core/vdaf.rs`.

File: core/src/vdaf.rs
Comment (MxmUrw) [reply core/src/vdaf.rs]: Thanks for your branch!

File: docs/samples/tasks.yaml
Comment (MxmUrw) [reply docs/samples/tasks.yaml]: We removed our changes.

File: core/src/dp.rs
Comment (MxmUrw) [reply core/src/dp.rs]: `DpStrategyInstance` is removed now.

**Comment by MxmUrw** [comment]
It seems like we have addressed all comments! What's next?

**Review by tgeoghegan — APPROVED** [review]
This LGTM. Next steps: I've tagged David to re-review in case he wants to take another look. David, if these changes are OK for you, then we should merge.

**Review by divergentdave — APPROVED** [review]
Looks good to me as well. Thank you for your patience!

**Comment by MxmUrw** [comment]
Very nice, thanks to everyone for merging!

**Pull request merged**
